canonical/ubuntu.com
# Engage pages pagination

## Done

- I've made some filters fixed, for example language filters have to be fixed, because if an author adds a new language, we need a thank-you template translated into the language, so it can't be used right after immediately, this forces authors to go through us first. Same with resource types, there should be a limited amount of them, otherwise authors may create things in the wrong place, case in point blogs, blog should be in blogs not in engage pages.
- Very few options around querying strings inside markdown (since this is not separate Postgres database field), make it very difficult to get a list of tags. So for now I just reused the query engage pages param data explorer query to create a list, which is not very performant, could be improved if anybody has better ideas.
- As a result, I also had to limit the filters to search one at a time, this is probably the fastest. If users ping us to ask for hierarchical search, this should be defined more clearly and we can write another data explorer query.
- I didn't add pagination for takeovers.json, as it is usually pretty low numbers, and it can be a bit more logic that needs to be discussed e.g. how many do we limit? how will stakeholders want to show this priority etc. But it is already supported, just need to change the view function to do it.


## QA

1. Go to https://ubuntu-com-14035.demos.haus/engage
2. Scroll down and click next page and/or the numbers to check that next page doesn't repeat the same engage pages.
3. Select language, and check that the correct language of engage pages shows
4. Same for resource types and tags

Check https://ubuntu-com-14035.demos.haus/tutorials work and check https://ubuntu-com-14035.demos.haus/takeovers work.

## Issue / Card

Fixes https://warthogs.atlassian.net/browse/WD-12662

## Help

[QA steps](https://discourse.canonical.com/t/qa-steps/152) - [Commit guidelines](https://discourse.canonical.com/t/commit-guidelines/148)
